"Polyfills" section: comparison table. Closes #111

catamphetamine opened the pull request:
* Created a dedicated "Polyfills" section in the readme.
* Added some clarification notes.
* Added a comparison table to avoid confusion.

Closes #111

longlho commented:
Just some context we used test262 to be completely spec-compliant which is why node 10+ is needed for formatToParts. Otherwise polyfilling that requires numberformat polyfills as well

catamphetamine commented:
@longlho So can I replace `Node 10+` in the "Requirements" with `Intl.NumberFormat`?

catamphetamine commented:
@longlho We could also add a "test262" section to the table if you want. I don't know what "test262" is so you'd have to phrase it for us.

littledan commented:
Do you know anything about why the bundle size of one of them is bigger?

catamphetamine commented:
@littledan I used `unpkg.com` to get the bundle sizes:
https://unpkg.com/intl-relative-time-format@1.0.6/dist/
I guess that the bundle for `intl-relative-time-format` is big because Frederik Wessberg @wessberg (the author) didn't include a "core" bundle in the distribution but instead put all locales into the bundle.
It's speculation though, perhaps @wessberg could answer in more details.

wessberg commented:
@catamphetamine, that is incorrect. No locale data is loaded by default.

catamphetamine commented:
@wessberg Ok, then the bundle size is big for some other reason.

wessberg commented:
@capamphetamine, I would suggest that we instead show sizes with minification since `intl-relative-time-format` is distributed without minification and with comments. Would lead to a more balanced comparison. 🙂

catamphetamine commented:
@wessberg Oh, so it's not minified. Perhaps you could minify it to measure the resulting core bundle size? Then we could add the correct size to the comparison table.

wessberg commented:
It's 12.33kb minified with Terser.

catamphetamine commented:
@wessberg Ok, I've pushed the updated bundle size to the PR.

wessberg commented:
Perfect 👍

longlho commented:
@littledan can provide more details on test262 but it’s the test suite used by spec implementers like v8 & spidermonkey to be fully spec-compliant (although it is subject to cldr version). There’re subtle impl differences like you can call Intl.RelativeTimeFormat without “new”, or -0 is “0 seconds ago” vs +0 is “in 0 seconds”. You can add that as a row as well.

catamphetamine commented:
@longlho 

> There’re subtle impl differences like you can call Intl.RelativeTimeFormat without “new”.

I guess this one is insignificant so it can be omitted.

> or -0 is “0 seconds ago” vs +0 is “in 0 seconds”. You can add that as a row as well.

A quote from http://2ality.com/2012/03/signedzero.html :

> It is claimed that the inclusion of signed zero in IEEE 754 makes it much easier to achieve numerical accuracy in some critical problems, in particular when computing with complex elementary functions. 

I'd argue that `-0` vs `+0` feature while easy to add is unnecessary for the task of formatting time because both "in 0 seconds" and "0 seconds ago" are actually "right now" and since formatting time is all about the common sense of a common human it shouldn't involve any purely mathematical concepts.
So I'd skip this mentioning "feature" as well.

The comparison table should be kept minimal only showcasing significant differences when using the library so that it doesn't add unnecessary confusion with `-0 vs +0` and similar purely technical details a generic library user might not be familiar with (and doesn't need to).

 @longlho If you have anything really significant for the end user to add to the PR post your suggestions in the form of the already formulated text. Otherwise the PR can be merged.

longlho commented:
That’s not correct. If you set numeric to always, what are the output of 0 vs -0 vs +0? Those are just examples, there are others surrounding how the prototype object should look like to be spec-compliant.

The “new” keywords affect caching infrastructure semantics (e.g fast-memoize which is super popular only cache function calls).

longlho commented:
It is subject to cldr version thou, e.g https://github.com/tc39/test262/pull/2191

catamphetamine commented:
@longlho 
I see, so basically what you're telling is that there's a test suite called `tc39/test262` that contains various test cases for a 100% compliant implementation of `Intl.RelativeTimeFormat`.
Ok then, I guess it could be added as a row in the comparison table.
Something like:

`Official ECMAScript Conformance Test | empty | empty | checkmark icon`

@longlho Would that be it?

longlho commented:
Yeah i think so

On Wed, Jun 19, 2019 at 12:46 PM Nikolay <notifications@github.com> wrote:

> @longlho <https://github.com/longlho>
> I see, so basically what you're telling is that there's a test suite
> called tc39/test262 that contains various test cases for a 100% compliant
> implementation of Intl.RelativeTimeFormat.
> Ok then, I guess it could be added as a row in the comparison table.
> Something like:
>
> Official ECMAScript Conformance Test | empty | empty | checkmark icon
>
> @longlho <https://github.com/longlho> Would that be it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tc39/proposal-intl-relative-time/pull/112?email_source=notifications&email_token=AABQM36J2BVKLJ7ZD3JJSTDP3JPG3A5CNFSM4HZKOGA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYCPCIY#issuecomment-503640355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABQM35KXFI6PZLN4ZJ3HILP3JPG3ANCNFSM4HZKOGAQ>
> .
>

catamphetamine commented:
Updated the PR with the new "Passes the Official ECMAScript Conformance Test" comparison table row.

Also removed the sequential polyfills list in favour of the just the comparison table.

wessberg commented:
If we're going to add the test262 conformance data to the comparison table, I think it needs to link to a source showing the test data results using each polyfill.

catamphetamine commented:
@longlho If you have the results of passing the `test262` suite then you can provide a link here so that it could be added to the table.

wessberg commented:
@catamphetamine, yes, that would be valuable, but in order for it to be a proper comparison, we should provide test results for _all_ 3 polyfills including raw data 🙂

longlho commented:
No I had to port the tests over to the formatjs repo. Test262-harness rn
doesn’t include relativetimeformat

On Wed, Jun 19, 2019 at 1:27 PM Frederik Wessberg <notifications@github.com>
wrote:

> @catamphetamine <https://github.com/catamphetamine>, yes, that would be
> valuable, but in order for it to be a proper comparison, we should provide
> test results for *all* 3 polyfills including raw data 🙂
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tc39/proposal-intl-relative-time/pull/112?email_source=notifications&email_token=AABQM35RLVJUS5CMVZKN663P3JUB3A5CNFSM4HZKOGA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYCSVLQ#issuecomment-503655086>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABQM35UHWUWBUQCMKUTK6LP3JUB3ANCNFSM4HZKOGAQ>
> .
>

wessberg commented:
In all fairness, and just to reiterate, I strongly suggest we either provide raw data based on the same test suite for all 3 polyfills, or leave the spec conformance data out entirely from the table.

catamphetamine commented:
@wessberg Oooh, I see what you mean, an ultimate test results comparison table.
Well, I guess it could be implemented as some generic "test suite" that would import each polyfill and perform the tests. Something like this:

```js
import polyfill1 from '...'
import polyfill2 from '...'
import polyfill3 from '...'

const RTF1 = polyfill1
const RTF2 = polyfill2
const RTF3 = polyfill3

function test(RTF) {
  // Perform the tests, generate the summary output.
}

test(RTF1)
test(RTF2)
test(RTF3)
```

@longlho Since you're already familiar with this thing and have made it running can you write the ultimate comparison script?
Otherwise it wouldn't be too fair to all the participants, I agree.
I'll remove the `Test262` row from the comparison table for now.

longlho commented:
I don't mind either way but won't have to time make the test suite work across implementations.

catamphetamine commented:
@longlho Or maybe you could publish a repo with the test runner and then each of the polyfill authors could fork that repo to run the tests against their library, so that it would be like a "standard" test suite for any `Intl.RelativeTimeFormat` polyfill.

catamphetamine commented:
@littledan I guess the `Test262` thing is gonna take some time due to the complexity of the whole thing, don't want to rush @longlho on this. Feel free to merge this PR as-is at any time if you feel like it: we can always create another one in the future when we have `Test262` suite properly set up and running.

littledan commented:
I'm going to wait on merging this PR until the authors of the referenced polyfills are happy with it.

catamphetamine commented:
I see, it won't be merged then. Too little cooperation.
No one's happy with the [previous PR](https://github.com/tc39/proposal-intl-relative-time/pull/110) that caused the confusion to begin with.

longlho commented:
That’s subjective

On Sat, Jun 22, 2019 at 6:14 AM Nikolay <notifications@github.com> wrote:

> I see, it won't be merged then. Too little cooperation.
> No one's happy with the previous PR
> <https://github.com/tc39/proposal-intl-relative-time/pull/110> that
> caused the confusion to begin with.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tc39/proposal-intl-relative-time/pull/112?email_source=notifications&email_token=AABQM32VDJD3GRZKIZBWPRLP3X3P3A5CNFSM4HZKOGA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYKGCXY#issuecomment-504652127>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABQM34VA54WRVQD7TGGDYDP3X3P3ANCNFSM4HZKOGAQ>
> .
>

longlho commented:
```bash
@formatjs/intl-relativetimeformat: Running "test262-harness --reporter-keys file,attrs,result -t 3 --prelude ./dist/polyfill-with-locales.js -r json /home/longlho/formatjs/test262/test/intl402/RelativeTimeFormat/**/*.js"
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ "RelativeTimeFormat" property of Intl.
@formatjs/intl-relativetimeformat: ✓ "RelativeTimeFormat" property of Intl.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ The prototype of the Intl.RelativeTimeFormat constructor is %FunctionPrototype%.
@formatjs/intl-relativetimeformat: ✓ Intl.RelativeTimeFormat instance object extensibility
@formatjs/intl-relativetimeformat: ✓ The prototype of the Intl.RelativeTimeFormat constructor is %FunctionPrototype%. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Intl.RelativeTimeFormat instance object extensibility
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Intl.RelativeTimeFormat instance object is created from %RelativeTimeFormatPrototype%.
@formatjs/intl-relativetimeformat: ✓ Intl.RelativeTimeFormat instance object is created from %RelativeTimeFormatPrototype%.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "prototype" property of the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks error cases for the locales argument to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks the "prototype" property of the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks error cases for the locales argument to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks various cases for the locales argument to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks various cases for the locales argument to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Verifies the NewTarget check for Intl.RelativeTimeFormat.
@formatjs/intl-relativetimeformat: ✓ Verifies the NewTarget check for Intl.RelativeTimeFormat.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of a null options argument to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks handling of a null options argument to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid value for the localeMatcher option to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid value for the localeMatcher option to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid value for the numeric option to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks handling of valid values for the numeric option to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid value for the numeric option to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the order of operations on the options argument to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks the order of operations on the options argument to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of valid values for the numeric option to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks that the RelativeTimeFormat constructor does not cause the
@formatjs/intl-relativetimeformat: NumberFormat and PluralRules constructors to get properties off
@formatjs/intl-relativetimeformat: Object.prototype through the options objects it creates.
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid value for the style option to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks that the RelativeTimeFormat constructor does not cause the
@formatjs/intl-relativetimeformat: NumberFormat and PluralRules constructors to get properties off
@formatjs/intl-relativetimeformat: Object.prototype through the options objects it creates.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid value for the style option to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of valid values for the style option to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of valid values for the style option to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks the propagation of exceptions from the options for the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks the propagation of exceptions from the options for the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor.
@formatjs/intl-relativetimeformat: ✓ Checks that RelativeTimeFormat can be subclassed.
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks that RelativeTimeFormat can be subclassed. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Tests that Intl.RelativeTimeFormat has a supportedLocalesOf property, and it works as planned.
@formatjs/intl-relativetimeformat: ✓ Tests that Intl.RelativeTimeFormat has a supportedLocalesOf property, and it works as planned.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Verifies there's no branding check for Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat: ✓ Verifies there's no branding check for Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks error cases for the locales argument to the supportedLocalesOf function. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks error cases for the locales argument to the supportedLocalesOf function.
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid values for the localeMatcher option to the supportedLocalesOf function.
@formatjs/intl-relativetimeformat: ✓ Checks handling of invalid values for the localeMatcher option to the supportedLocalesOf function. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of a null options argument to the supportedLocalesOf function. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of a null options argument to the supportedLocalesOf function.
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object options arguments to the supportedLocalesOf function.
@formatjs/intl-relativetimeformat: ✓ Checks handling of an undefined options argument to the supportedLocalesOf function.
@formatjs/intl-relativetimeformat: ✓ Checks handling of non-object options arguments to the supportedLocalesOf function. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "supportedLocalesOf" property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks handling of an undefined options argument to the supportedLocalesOf function. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "supportedLocalesOf" property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Verifies the type of the return value of Intl.RelativeTimeFormat.supportedLocalesOf().
@formatjs/intl-relativetimeformat: ✓ Verifies the type of the return value of Intl.RelativeTimeFormat.supportedLocalesOf(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "constructor" property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Checks the "constructor" property of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Verifies the branding check for the "format" function of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Verifies the branding check for the "format" function of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English.
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "format" property of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "format" property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Verifies the branding check for the "formatToParts" function of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Verifies the branding check for the "formatToParts" function of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English.
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
@formatjs/intl-relativetimeformat: ✓ Checks the "formatToParts" property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Checks the "formatToParts" property of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format().
@formatjs/intl-relativetimeformat: ✓ Verifies the branding check for the "resolvedOptions" function of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Verifies the branding check for the "resolvedOptions" function of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Verifies that the return value of Intl.RelativeTimeFormat.prototype.resolvedOptions() is not cached.
@formatjs/intl-relativetimeformat: ✓ Verifies that the return value of Intl.RelativeTimeFormat.prototype.resolvedOptions() is not cached. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.resolvedOptions(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.resolvedOptions().
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.resolvedOptions().
@formatjs/intl-relativetimeformat: ✓ Verifies the property order for the object returned by resolvedOptions().
@formatjs/intl-relativetimeformat: ✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.resolvedOptions(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Verifies the property order for the object returned by resolvedOptions(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "resolvedOptions" property of the RelativeTimeFormat prototype object. (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the "resolvedOptions" property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Checks the properties of the result of Intl.RelativeTimeFormat.prototype.resolvedOptions().
@formatjs/intl-relativetimeformat: ✓ Checks Object.prototype.toString with Intl.RelativeTimeFormat objects.
@formatjs/intl-relativetimeformat: ✓ Checks the properties of the result of Intl.RelativeTimeFormat.prototype.resolvedOptions(). (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks Object.prototype.toString with Intl.RelativeTimeFormat objects.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: ✓ Checks the @@toStringTag property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat: ✓ Checks the @@toStringTag property of the RelativeTimeFormat prototype object.
@formatjs/intl-relativetimeformat:  (Strict Mode)
@formatjs/intl-relativetimeformat: Tests: 150 passed, 150 total
```
You guys can follow something similar to this PR https://github.com/eemeli/intl-pluralrules/pull/9

wessberg commented:
I’ve integrated the `test262-harness` submodule with [`intl-relative-time-format`](http://github.com/wessberg/intl-relative-time-format/) and here’s the test results: (150/150 passing tests):

```
Running "test262-harness --hostArgs "--icu-data-dir=node_modules/full-icu" --reporter-keys file,attrs,result -t 8 --prelude ./test262-polyfill.js -r json test262/test/intl402/RelativeTimeFormat/**/*.*"
✓ Checks the "length" property of the RelativeTimeFormat constructor.

✓ The prototype of the Intl.RelativeTimeFormat constructor is %FunctionPrototype%. (Strict Mode)
✓ "RelativeTimeFormat" property of Intl.
 (Strict Mode)
✓ The prototype of the Intl.RelativeTimeFormat constructor is %FunctionPrototype%.
✓ "RelativeTimeFormat" property of Intl.

✓ Checks the "name" property of the RelativeTimeFormat constructor.

✓ Checks the "name" property of the RelativeTimeFormat constructor.
 (Strict Mode)
✓ Checks the "length" property of the RelativeTimeFormat constructor.
 (Strict Mode)
✓ Checks the "prototype" property of the RelativeTimeFormat constructor.

✓ Intl.RelativeTimeFormat instance object extensibility
 (Strict Mode)
✓ Intl.RelativeTimeFormat instance object is created from %RelativeTimeFormatPrototype%.

✓ Checks error cases for the locales argument to the RelativeTimeFormat constructor. (Strict Mode)
✓ Intl.RelativeTimeFormat instance object extensibility

✓ Checks error cases for the locales argument to the RelativeTimeFormat constructor.
✓ Checks the "prototype" property of the RelativeTimeFormat constructor.
 (Strict Mode)
✓ Intl.RelativeTimeFormat instance object is created from %RelativeTimeFormatPrototype%.
 (Strict Mode)
✓ Verifies the NewTarget check for Intl.RelativeTimeFormat.
 (Strict Mode)
✓ Checks handling of a null options argument to the RelativeTimeFormat constructor. (Strict Mode)
✓ Verifies the NewTarget check for Intl.RelativeTimeFormat.

✓ Checks handling of a null options argument to the RelativeTimeFormat constructor.
✓ Checks handling of invalid value for the localeMatcher option to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks various cases for the locales argument to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of invalid value for the localeMatcher option to the RelativeTimeFormat constructor.
✓ Checks various cases for the locales argument to the RelativeTimeFormat constructor.
✓ Checks handling of invalid value for the numeric option to the RelativeTimeFormat constructor.
✓ Checks handling of valid values for the numeric option to the RelativeTimeFormat constructor.
✓ Checks handling of valid values for the numeric option to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of invalid value for the numeric option to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks that the RelativeTimeFormat constructor does not cause the
NumberFormat and PluralRules constructors to get properties off
Object.prototype through the options objects it creates.
 (Strict Mode)
✓ Checks that the RelativeTimeFormat constructor does not cause the
NumberFormat and PluralRules constructors to get properties off
Object.prototype through the options objects it creates.

✓ Checks the order of operations on the options argument to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks the order of operations on the options argument to the RelativeTimeFormat constructor.
✓ Checks handling of invalid value for the style option to the RelativeTimeFormat constructor.
✓ Checks the propagation of exceptions from the options for the RelativeTimeFormat constructor.
✓ Checks the propagation of exceptions from the options for the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of invalid value for the style option to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor.
✓ Checks handling of valid values for the style option to the RelativeTimeFormat constructor.
✓ Checks handling of valid values for the style option to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor.
✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor. (Strict Mode)
✓ Tests that Intl.RelativeTimeFormat has a supportedLocalesOf property, and it works as planned.

✓ Tests that Intl.RelativeTimeFormat has a supportedLocalesOf property, and it works as planned.
 (Strict Mode)
✓ Checks that RelativeTimeFormat can be subclassed.
✓ Checks that RelativeTimeFormat can be subclassed. (Strict Mode)
✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor. (Strict Mode)
✓ Checks handling of non-object option arguments to the RelativeTimeFormat constructor.
✓ Verifies there's no branding check for Intl.RelativeTimeFormat.supportedLocalesOf().
 (Strict Mode)
✓ Verifies there's no branding check for Intl.RelativeTimeFormat.supportedLocalesOf().

✓ Checks error cases for the locales argument to the supportedLocalesOf function. (Strict Mode)
✓ Checks the "name" property of Intl.RelativeTimeFormat.supportedLocalesOf().

✓ Checks error cases for the locales argument to the supportedLocalesOf function.
✓ Checks the "name" property of Intl.RelativeTimeFormat.supportedLocalesOf().
 (Strict Mode)
✓ Checks the "length" property of Intl.RelativeTimeFormat.supportedLocalesOf().

✓ Checks the "length" property of Intl.RelativeTimeFormat.supportedLocalesOf().
 (Strict Mode)
✓ Checks handling of invalid values for the localeMatcher option to the supportedLocalesOf function. (Strict Mode)
✓ Checks handling of invalid values for the localeMatcher option to the supportedLocalesOf function.
✓ Checks handling of an undefined options argument to the supportedLocalesOf function. (Strict Mode)
✓ Checks handling of an undefined options argument to the supportedLocalesOf function.
✓ Checks handling of non-object options arguments to the supportedLocalesOf function. (Strict Mode)
✓ Checks handling of non-object options arguments to the supportedLocalesOf function.
✓ Checks handling of a null options argument to the supportedLocalesOf function. (Strict Mode)
✓ Checks handling of a null options argument to the supportedLocalesOf function.
✓ Checks the "supportedLocalesOf" property of the RelativeTimeFormat prototype object.
 (Strict Mode)
✓ Verifies the type of the return value of Intl.RelativeTimeFormat.supportedLocalesOf(). (Strict Mode)
✓ Checks the "constructor" property of the RelativeTimeFormat prototype object.
✓ Checks the "supportedLocalesOf" property of the RelativeTimeFormat prototype object.

✓ Verifies the branding check for the "format" function of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Verifies the branding check for the "format" function of the RelativeTimeFormat prototype object.
✓ Checks the "constructor" property of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Verifies the type of the return value of Intl.RelativeTimeFormat.supportedLocalesOf().
✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.format().
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English. (Strict Mode)
✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in English.
✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.format().
✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
✓ Checks the "format" property of the RelativeTimeFormat prototype object.
✓ Checks the "format" property of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.format().
✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format().
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.format().
✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format().
✓ Verifies the branding check for the "formatToParts" function of the RelativeTimeFormat prototype object.
✓ Verifies the branding check for the "formatToParts" function of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English.
✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format().
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English. (Strict Mode)
✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.formatToParts() in English. (Strict Mode)
✓ Checks the "formatToParts" property of the RelativeTimeFormat prototype object.
✓ Checks the "formatToParts" property of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish. (Strict Mode)
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
✓ Checks the behavior of Intl.RelativeTimeFormat.prototype.format() in Polish.
✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Checks the handling of plural unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Checks the handling of invalid unit arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts().
✓ Verifies the branding check for the "resolvedOptions" function of the RelativeTimeFormat prototype object.
✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.resolvedOptions().
✓ Checks the handling of invalid value arguments to Intl.RelativeTimeFormat.prototype.formatToParts(). (Strict Mode)
✓ Verifies the branding check for the "resolvedOptions" function of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Verifies that the return value of Intl.RelativeTimeFormat.prototype.resolvedOptions() is not cached. (Strict Mode)
✓ Verifies that the return value of Intl.RelativeTimeFormat.prototype.resolvedOptions() is not cached.
✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format().
✓ Checks the handling of non-number value arguments to Intl.RelativeTimeFormat.prototype.format(). (Strict Mode)
✓ Checks the "length" property of Intl.RelativeTimeFormat.prototype.resolvedOptions(). (Strict Mode)
✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.resolvedOptions().
✓ Checks the "name" property of Intl.RelativeTimeFormat.prototype.resolvedOptions(). (Strict Mode)
✓ Checks the properties of the result of Intl.RelativeTimeFormat.prototype.resolvedOptions().
✓ Verifies the property order for the object returned by resolvedOptions(). (Strict Mode)
✓ Verifies the property order for the object returned by resolvedOptions().
✓ Checks the "resolvedOptions" property of the RelativeTimeFormat prototype object. (Strict Mode)
✓ Checks the "resolvedOptions" property of the RelativeTimeFormat prototype object.
✓ Checks the @@toStringTag property of the RelativeTimeFormat prototype object.

✓ Checks the @@toStringTag property of the RelativeTimeFormat prototype object.
 (Strict Mode)
✓ Checks Object.prototype.toString with Intl.RelativeTimeFormat objects.

✓ Checks Object.prototype.toString with Intl.RelativeTimeFormat objects.
 (Strict Mode)
✓ Checks the properties of the result of Intl.RelativeTimeFormat.prototype.resolvedOptions(). (Strict Mode)
Tests: 150 passed, 150 total
```

catamphetamine commented:
@longlho @wessberg I see, so I'm the last one...
I guess I'll pass on implementing all of the 150 test cases, [about 100](https://github.com/catamphetamine/Intl.RelativeTimeFormat-test262) is fine for me.
I've updated the table to reflect the Test262 passage status.
I've also moved my library to the right side after yours because it chose to not implement the full list of the test cases.
See if you're ok with the results.

longlho commented:
u can actually replace the size w/ badge since it's live:

![intl-relative-time-format](https://badgen.net/bundlephobia/minzip/intl-relative-time-format?label=intl-relative-time-format)
![relative-time-format](https://badgen.net/bundlephobia/minzip/relative-time-format?label=relative-time-format)
![@formatjs/intl-relativetimeformat](https://badgen.net/bundlephobia/minzip/@formatjs/intl-relativetimeformat?label=@formatjs/intl-relativetimeformat)

wessberg commented:
LGTM

catamphetamine commented:
@longlho I see, so it downloads a package, runs `npm install`, passes `${packagePath}/index.js` to webpack and returns the bundle size after `gzipSync()`.
https://github.com/pastelsky/package-build-stats/blob/master/src/getPackageStats.js

The images you have provided have non-human-readable URLs so it's unclear whether they're stale or do they refresh themselves.
I found another service for that called ["bundlephobia"](https://shields.io/category/size) which seems to calculate the sizes dynamically.
I've added these badges to the table instead of the actual sizes.
@longlho See if it works.
https://github.com/catamphetamine/proposal-intl-relative-time/blob/patch-1/README.md#polyfills

longlho commented:
the badges I listed were bundlephobia so that works :)

catamphetamine commented:
@littledan Alright, looks like we finally made it.

littledan commented:
OK, sounds like the authors all agree on this comparison table, so I'm merging it. Please let me know if there are issues remaining.
